worlds/pokemon_crystal/phone.py

```diff
@@ -0,0 +1,81 @@
+import random
+from typing import TYPE_CHECKING
+
+from BaseClasses import ItemClassification
+from .data import data
+from .phone_data import get_shuffled_basic_calls, template_call_bike_shop, template_call_psychic, template_call_remote
+
+if TYPE_CHECKING:
+    from . import PokemonCrystalWorld
+else:
+    PokemonCrystalWorld = object
+
+
+def generate_phone_traps(world: PokemonCrystalWorld):
+    if world.options.phone_trap_weight.value:
+        bike_shop_location = None
+        psychic_location = None
+        remote_progression_locs = []
+        for location in world.multiworld.get_locations(world.player):
+            if location.name == data.locations["BICYCLE"].label:
+                # world.multiworld.worlds[world.player].game
+                bike_shop_location = location
+            elif location.name == data.locations["TM29_PSYCHIC"].label:
+                psychic_location = location
+            elif location.address is not None and location.item and location.item.player != world.player and location.item.classification == ItemClassification.progression:
+                remote_progression_locs.append(location)
+
+        remote_progression_spheres = []
+        sphere_i = 0
+        for sphere in world.multiworld.get_spheres():
```

Comment:
world.multiworld.get_spheres() is computationally expensive and **not cached**.

If you don't do anything that modifies spheres (which generally, you really shouldn't if you can help it), you should grab them once in a `stage_` function as a class variable, and share them between world instances

Reply:
Oh. Lmao

Reply:
<img width="921" alt="image" src="https://github.com/ArchipelagoMW/Archipelago/assets/75546613/5fa4017c-c6f7-4d97-8e40-c2b3621ee88a">
What timing lol